src/main/scala/io/findify/flinkadt/api/serializer/ScalaCaseClassSerializer.scala

```diff
@@ -0,0 +1,88 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.findify.flinkadt.api.serializer
+
+import io.findify.flinkadt.api.serializer.ScalaCaseClassSerializer.lookupConstructor
+import org.apache.flink.api.common.typeutils.{TypeSerializer, TypeSerializerSnapshot}
+
+import java.io.ObjectInputStream
+
+import scala.annotation.nowarn
+import scala.reflect.runtime.universe
+
+/** This is a non macro-generated, concrete Scala case class serializer.
+  * Copied from Flink 1.14 without `SelfResolvingTypeSerializer`.
+  */
+@SerialVersionUID(1L)
+class ScalaCaseClassSerializer[T <: Product](
+    clazz: Class[T],
+    scalaFieldSerializers: Array[TypeSerializer[_]]
+) extends CaseClassSerializer[T](clazz, scalaFieldSerializers) {
+
+  @transient
+  private var constructor = lookupConstructor(clazz)
+
+  override def createInstance(fields: Array[AnyRef]): T = {
+    constructor(fields)
+  }
+
+  override def snapshotConfiguration(): TypeSerializerSnapshot[T] =
+    new ScalaCaseClassSerializerSnapshot[T](this)
+
+  // Do NOT delete this method, it is used by ser/de even though it is private.
+  // This should be removed once we make sure that serializer are no long java serialized.
+  private def readObject(in: ObjectInputStream): Unit = {
+    in.defaultReadObject()
+    constructor = lookupConstructor(clazz)
+  }
+}
+
+object ScalaCaseClassSerializer {
+  @nowarn("msg=(eliminated by erasure)|(explicit array)")
+  def lookupConstructor[T](cls: Class[T]): Array[AnyRef] => T = {
```

Comment:
This is from flink-scala and admittedly it's pretty jank. It should be okay to grab `rawConstruct` from magnolia and pass that to the serializer instead.

Reply:
Following up on my own comment - flink serializers & serializer snapshotters are really annoying to work with and have pretty strict requirements. I.e. if you pass a function like `Array[AnyRef] => A` as a parameter to a serializer (for say supporting product types / case classes), you are also responsible for serializing & deserializing that function with the snapshotter. This is why the constructor is looked up using the class instead - the snapshotter has the capability of serializing class names, then deserializing them later and doing a lookup via a provided class loader. 
Supporting this behavior in Scala 3 will be tricky, since Scala 3 drops the official runtime reflection support, instead only providing compile-time macros.